README.md
MrTarget by Renevatium

https://mods.curse.com/addons/wow/mrtarget


MrTarget/CHANGELOG
MrTarget v5.1.0
=====================================================================
Copyright (C) 2016 Lock of War, Renevatium

Warlords of Draenor v7.1.5
---------------------------------------------------------------------
13/01/2017
v5.1.0
Updated to support patch 7.1.5 (70100)
Fixed for updated GetSpecializationInfoForClassID

Warlords of Draenor v7.1.0
---------------------------------------------------------------------
05/11/2016
v5.0.1
Fixes to interfaces options
Use map ID to identify battlegrounds
Use 10man config for RBG

26/10/2016
v5.0.0
Updated to support patch 7.1.0 (70100)
Fixed player target icon bug
Updated range tracking

Warlords of Draenor v7.0.3
---------------------------------------------------------------------
25/10/2016
v4.1.0
Fixed not loading in battlegrounds
Per battlefield configuration
Fixed non-existant target error
Prepared for patch 7.1

25/09/2016
v4.0.1
Fixed bug causing frames to appear randomly (shol)
Friendly units option displaying correctly
Per character configuration (aaron)
Persistant configuration
Option to hide enemy frames
Preparation for 40man
Updated copyright text

02/08/2016
v4.0.0
Added columns option
Added Grid layout
Full fade when out of range
Fixed range logic

21/07/2016
v3.1.5
Updated to support patch 7.0.3 (70000)

Warlords of Draenor v6.2.4
---------------------------------------------------------------------
16/04/2016
v3.1.4
Updated to support patch 6.2.4 (60200)
Fixed error from GetSpecializationInfo

Warlords of Draenor v6.2.3
---------------------------------------------------------------------
18/11/2015
v3.1.3
Updated to support patch 6.2.3 (60200)
Party frames no longer shown by default
Auras now obey orientation
Disabled for Isle of Conquest and Alterac Valley

Warlords of Draenor v6.2.2
---------------------------------------------------------------------
07/09/2015
v3.1.2
Updated to support patch 6.2.2 (60200)
Fix hide/show slash commands

Warlords of Draenor v6.2.1
---------------------------------------------------------------------
21/08/2015
v3.1.1
Updated to support patch 6.2.1 (60200)
Fix range tracking
Fixed objectives frame toggle

Warlords of Draenor v6.2.0
---------------------------------------------------------------------
10/08/2015
v3.1.0
Party frames
Dropped arena auras

23/07/2015
v3.0.1
Improved targeting
Improved unit detection
Switch from UNIT_HEALTH to UNIT_HEALTH_FREQUENT event
Added UPDATE_MOUSEOVER_UNIT event

17/07/2015
v3.0.0
Total rebuild for patch 6.2.0 (60200)
Dropepd arena functionality (New arena addon coming soon)
New raid frames (party frames for HEALERS coming soon)
New unit frames
New aura system
New range detection system
Insane performance improvement
Upgraded UI elements

Warlords of Draenor v6.1.2
---------------------------------------------------------------------
16/04/2015
v2.2.3
Fixed nil name error in Arena
Fixed multiple issues with Arena loading

06/04/2015
v2.2.2
Fixed missing buff icons

02/04/2015
v2.2.1
Fixed hide/show /commands
Fixed CompactRaidManager bug (jbbuena)
Added player cast debuffs in BGs (teakan)
Added option to hide power bar (teakan)
Updated for patch 6.1.2 (60100)

Warlords of Draenor v6.1.0
---------------------------------------------------------------------
25/02/2015
v2.1.4
Updated for patch 6.1.0 (60100)
Fixed targeted count (SantiagoJH90)
Generally fixed lots of bugs

02/03/2015
v2.1.5
Fixed first load bug (seatleson, evilmonkeyz)
Improved range tracking

25/02/2015
v2.1.4
Updated for patch 6.1.0 (60100)
Fixed targeted count (SantiagoJH90)
Generally fixed lots of bugs

Warlords of Draenor v6.0.3
---------------------------------------------------------------------
18/02/2015
v2.1.3
Fixed font on aura timers in borderless
Added friendly target counters (SantiagoJH90)

09/02/2015
v2.1.2
Fixed several range issues
Added transmute naming option (pokemon only)
Disable naming option during battle
Improved health tracking
Added spec icons to Borderless UI

23/01/2015
v2.1.1
Added Russian transliteration (Sijtsma)
Only movable with options panel
Split battleground and arena settings
Borderless/minimalist/shadowed ui interface option (deathmonkey)
Fixed options saving bug
Added range tracking
Fixed unit frame depth (Kaligma)
Significantly improved performance

11/12/2014
v2.0.4
Fixed bug causing objectives to hide in Garrisson

11/12/2014
v2.0.3
Added enable/disable option for battlefields
Added enable/disable option for renaming
Added right click /focus (Dazian)
Added hide/show slash commands
Fixed bug: not loading immediately in BG (Kelsey)
Fixed bug: clearing randomly (ed)
Objectives frame hides

01/12/2014
v2.0.2
Update arena auras
Removed objectives frame hide (taint)

30/11/2014
v2.0.1
Disabled range
Updated Arena auras

28/11/2014
v2.0.0
Now resizable
Added renaming sets
Added interface options panel
Updated aura (flag) recovery
Added Focused/Brutal assault icons
Arena system (beta)
Added Arena auras
GameTooltips for auras
28 yard range

27/10/2014
v1.1.3
Fixed issue with full team name replacement
Fixed auras [Deepwind Gorge]
Added slash command to enabled/disable renaming
Added specialization text to unit frames
Tooltips implemented but disabled

Warlords of Draenor v6.0.2
---------------------------------------------------------------------
15/10/2014
v1.1.2
Fix for ObjectiveTrackerFrame
Upgrade to Interface 60000 (WoD)

Mist of Pandaria v5.4.8
---------------------------------------------------------------------
10/10/2014
v1.1.1
Name ordering fixed permanently (sorry)
Arena build fixed [disabled]

08/10/2014
v1.1.0
Fixed incorrect ordering
Moved target/assist icons to the left
Enlarged and moved aura icons to the right
First Arena build [disabled]

09/09/2014
v1.0.4
Fixed Auctionator error (LP)


MrTarget/MrTargetAuras.lua
-- MrTargetAuras
-- =====================================================================
-- Copyright (C) 2016 Lock of War, Renevatium
--

local PVP_AURAS = {};
local PVP_AURAS_TEMP = {
  23333,23335,34976,46393,46392,141210,140876,156618,156621,121164,
  121175,121176,121177,125344,125345,125346,125347
};

for i=1, #PVP_AURAS_TEMP do
  local name, _, icon = GetSpellInfo(PVP_AURAS_TEMP[i]);
  if name then
    PVP_AURAS[i] = name;
  end
end

MrTargetAuras = {
  parent=nil,
  max=9,
  count=1
};

MrTargetAuras.__index = MrTargetAuras;

function MrTargetAuras:New(parent)
  local this = setmetatable({}, MrTargetAuras);
  this.frames = setmetatable({}, nil);
  this.auras = setmetatable({}, nil);
  this.parent = parent;
  for i=1, this.max do
    this.frames[i] = CreateFrame('Button', parent.frame:GetName()..'Aura'..i, parent.frame, 'MrTargetAuraTemplate');
    this.frames[i]:SetScript('OnUpdate', function(frame) this:OnUpdate(frame); end);
    this.frames[i]:Show();
  end
  return this;
end

function MrTargetAuras:SetAura(count, frame, id, name, duration, expires, icon)
  if not self.auras[id] then
    self.auras[id] = name;
    frame.id = id;
    frame.spell = name;
    frame.update = GetTime();
    frame.duration = duration;
    frame.time = expires-GetTime();
    frame.icon = icon;
    frame.ICON:SetTexture(icon);
    frame:ClearAllPoints();
    if self.parent.parent.reverse then frame:SetPoint('TOPRIGHT', frame:GetParent(), 'TOPLEFT', -4-((count-1)*38), 0);
    else frame:SetPoint('TOPLEFT', frame:GetParent(), 'TOPRIGHT', 4+((count-1)*38), 0);
    end
    return 1;
  end
  return 0;
end

function MrTargetAuras:UnsetAura(frame)
  frame.id = nil;
  frame.spell = nil;
  frame.update = GetTime();
  frame.duration = 0;
  frame.time = 0;
  frame.icon = nil;
  frame.ICON:SetTexture(nil);
  frame.expires:SetText('');
end

function MrTargetAuras:UnitAura(unit)
  if MrTarget:GetOption('AURAS') then
    self.auras = table.wipe(self.auras);
    self.count = self:UpdateCarriers(1, unit);
    self.count = self:UpdateDebuff(self.count, unit);
    for i=self.count,self.max do
      self:UnsetAura(self.frames[i]);
    end
  end
end

function MrTargetAuras:UpdateCarriers(count, unit)
  for i=1,#PVP_AURAS do
    if count > self.max then break; end
    local name, rank, icon, stack, type, duration, expires, source, _, _, id = UnitAura(unit, PVP_AURAS[i]);
    if name and icon then
      count = count+self:SetAura(count, self.frames[count], id, name, duration, expires, icon);
    end
  end
  return count;
end

function MrTargetAuras:UpdateDebuff(count, unit)
  for i=1,40 do
    if count > self.max then break; end
    local name, rank, icon, stack, type, duration, expires, source, _, _, id = UnitDebuff(unit, i, 'PLAYER');
    if name and icon then
      count = count+self:SetAura(count, self.frames[count], id, name, duration, expires, icon);
    end
  end
  return count;
end

function MrTargetAuras:UpdatePositions()
  local count = 1;
  for i=1,self.max do
    if self.frames[i].id then
      self.frames[i]:ClearAllPoints();
      if self.parent.parent.reverse then self.frames[i]:SetPoint('TOPRIGHT', self.frames[i]:GetParent(), 'TOPLEFT', -4-((count-1)*38), 0);
      else self.frames[i]:SetPoint('TOPLEFT', self.frames[i]:GetParent(), 'TOPRIGHT', 4+((count-1)*38), 0);
      end
      count = count+1;
    end
  end
end

function MrTargetAuras:Destroy()
  for i=1,self.max do
    self:UnsetAura(self.frames[i]);
  end
end

function MrTargetAuras:UpdateExpires(frame)
  if frame.time then
    frame.time = tonumber(frame.time) - (GetTime()-frame.update);
    frame.time = math.floor((frame.time*10)+0.5)/10;
    if frame.duration == 0 then
      frame.expires:SetText('');
    elseif frame.time < 0 then
      self:UnsetAura(frame);
      self:UpdatePositions();
    elseif frame.time <= 60 then
      local time = frame.time > 0 and frame.time or '';
      frame.expires:SetText(time);
    else
      local msg, val = SecondsToTimeAbbrev(frame.time);
      frame.expires:SetText(format(msg, val));
    end
  end
end

function MrTargetAuras:OnUpdate(frame)
  if frame.id then
    if GetTime()-frame.update >= 0.1 then
      self:UpdateExpires(frame);
      frame.ICON:SetTexture(frame.icon);
      frame.update = GetTime();
    end
  end
end

MrTarget/MrTargetRange.lua
-- MrTargetRange
-- =====================================================================
-- Copyright (C) 2016 Lock of War, Renevatium
--

MrTargetRange = {
  frame=nil,
  parent=nil,
  range=nil,
  update=0,
  frequency=0.25,
  harmful={},
  helpful={}
};

MrTargetRange.__index = MrTargetRange;

function MrTargetRange:New(parent)
  local this = setmetatable({}, MrTargetRange);
  this.harmful = setmetatable({}, nil);
  this.helpful = setmetatable({}, nil);
  this.parent = parent;
  this.frame = CreateFrame('Frame', parent.frame:GetName()..'Range', parent.frame);
  this.frame:SetScript('OnUpdate', function(frame, time) this:OnUpdate(time); end);
  this.frame:SetScript('OnEvent', function(frame, ...) this:OnEvent(...); end);
  this.frame:RegisterEvent('ACTIVE_TALENT_GROUP_CHANGED');
  -- this.frame:RegisterEvent('COMBAT_LOG_EVENT_UNFILTERED');
  this:UpdateSpells();
  this.frame:Show();
  return this;
end

local function SortByRange(u,v)
  if v and u then
    if u.range > v.range then
      return true;
    end
  elseif u then
    return true;
  end
end

function MrTargetRange:UpdateSpells()
  self.harmful = table.wipe(self.harmful);
  self.helpful = table.wipe(self.helpful);
  local numTabs = GetNumSpellTabs();
  for i=1,numTabs do
    local name,texture,offset,numSpells = GetSpellTabInfo(i);
    for j=1,numSpells do
      local id = j+offset;
      local name, rank = GetSpellBookItemName(id, 'spell');
      local range = select(6, GetSpellInfo(name));
      if range then
        if IsHarmfulSpell(id, 'spell') then
          table.insert(self.harmful, { name=name, range=range });
        elseif IsHelpfulSpell(id, 'spell') then
          table.insert(self.helpful, { name=name, range=range });
        end
      end
    end
  end
  table.sort(self.harmful, SortByRange)
  table.sort(self.helpful, SortByRange)
end

function MrTargetRange:GetHarmfulRange()
  local range = nil;
  for i=1, #self.harmful do
    if IsSpellInRange(self.harmful[i].name, self.parent.unit) == 1 then
      range = range == nil and self.harmful[i].range or math.min(range, self.harmful[i].range);
      if range then
        break;
      end
    end
  end
  return range;
end

function MrTargetRange:GetHelpfulRange()
  local range = nil;
  for i=1, #self.helpful do
    if IsSpellInRange(self.helpful[i].name, self.parent.unit) == 1 then
      range = range == nil and self.helpful[i].range or math.min(range, self.helpful[i].range);
      if range then
        break;
      end
    end
  end
  return range;
end

function MrTargetRange:OnUpdate(time)
  if UnitIsDeadOrGhost('player') then
    self.parent.range = nil;
    self.range = nil;
    self.update = 0;
    return;
  end
  self.update = self.update + time;
  if self.update > self.frequency then
    if self.parent:GetUnit(self.parent.unit) then
      if UnitIsConnected(self.parent.unit) and not UnitIsDeadOrGhost(self.parent.unit) then
        if UnitIsEnemy('player', self.parent.unit) then
          self.range = self:GetHarmfulRange();
        else
          self.range = self:GetHelpfulRange();
        end
      end
      self.parent.range = self.range;
      self.update = 0;
    end
  end
end

-- function MrTargetRange:CombatLogRangeCheck(sourceName, destName, spellId)
--   if sourceName and self.parent.name == sourceName then
--     if UnitIsEnemy('player', sourceName) then
--       print('Enemy in range '..sourceName);
--       self.range = self:GetHarmfulRange();
--       return;
--     else
--       return;
--     end
--   end
--   if destName and self.parent.name == destName then
--     if UnitIsEnemy('player', destName) then
--       print('Enemy in range '..destName);
--       self.range = self:GetHarmfulRange();
--       return;
--     else
--       return;
--     end
--   end
-- end

function MrTargetRange:OnEvent(event, unit, ...)
  if event == 'ACTIVE_TALENT_GROUP_CHANGED' then
    self:UpdateSpells();
  -- elseif event == 'COMBAT_LOG_EVENT_UNFILTERED' then
  --   local _, _, _, sourceName, _, _, _, destName, _, _, spellId = ...;
  --   self:CombatLogRangeCheck(sourceName, destName, spellId);
  end
end

MrTarget/MrTarget.lua
-- MrTarget v5.1.0
-- =====================================================================
-- This Work is provided under the Creative Commons
-- Attribution-NonCommercial-NoDerivatives 4.0 International Public License
--
-- Please send any bugs or feedback to mrtarget@lockofwar.com.
-- Debug /run print((select(4, GetBuildInfo())));

local BATTLEFIELD_SIZES = { 10, 15, 40 };

local DEFAULT_BATTLEFIELD_OPTIONS = {
  ENABLED=true,
  ENEMY=true,
  FRIENDLY=true,
  POSITION={
    HARMFUL={ 'TOPRIGHT', nil, 'TOPRIGHT', -(GetScreenWidth()*UIParent:GetEffectiveScale())/15, -(GetScreenHeight()*UIParent:GetEffectiveScale())/3 },
    HELPFUL={ 'TOPLEFT', nil, 'TOPLEFT', (GetScreenWidth()*UIParent:GetEffectiveScale())/15, -(GetScreenHeight()*UIParent:GetEffectiveScale())/3 },
  },
  BORDERLESS=false,
  ICONS=false,
  SIZE=100,
  NAMING='Transmute',
  POWER=true,
  RANGE=true,
  TARGETED=true,
  AURAS=true,
  COLUMNS=1
};

function copy(obj, seen)
  if type(obj) ~= 'table' then return obj end
  if seen and seen[obj] then return seen[obj] end
  local s = seen or {}
  local res = setmetatable({}, getmetatable(obj))
  s[obj] = res
  for k, v in pairs(obj) do res[copy(k, s)] = copy(v, s) end
  return res
end

local DEFAULT_OPTIONS = {}
for k,v in pairs(BATTLEFIELD_SIZES) do
  DEFAULT_OPTIONS[v] = copy(DEFAULT_BATTLEFIELD_OPTIONS);
end

DEFAULT_OPTIONS[15].COLUMNS = 2;
DEFAULT_OPTIONS[15].AURAS = false;
DEFAULT_OPTIONS[40].COLUMNS = 4;
DEFAULT_OPTIONS[40].FRIENDLY = false;
DEFAULT_OPTIONS[40].AURAS = false;

local FRIENDS = {
  'Jaina', 'Uther', 'Anduin', 'Valeera', 'Thrall', 'Gul\'dan', 'Garrosh', 'Arthas',
  'Malfurian', 'Rexxar', 'Magni', 'Alleria', 'Medivh', 'Chen', 'Antonidas', 'Sylvanas',
  'Muradin', 'Falstad', 'Alleria', 'Illidan', 'Tyrande', 'Gazlowe', 'Cairne', 'Draka',
  'Aegwynn', 'Arthas', 'Bolvar', 'Khadgar', 'Rhonin', 'Tirion', 'Varian', 'Brann',
  'Llane', 'Grommash', 'Orgrim', 'Durotan', 'Millhouse', 'Garona', 'Vol\'jin', 'Maleki'
};

local ENEMIES = {
  'Афила', 'Сэйбот', 'Яджун', 'Найнс', 'Супералёнка', 'Аллорион', 'Марги', 'Алёнаболт',
  'Скорозасияем', 'Вайлен', 'Лукертс', 'Игми', 'Вандерер', 'Биотикус', 'Эксдаркикс', 'Мельда',
  'Альвеона', 'Сигр', 'Альвеоняша', 'Айвен', 'Эрриган', 'Иллиняша', 'Алеандра', 'Атжай',
  'Койрэ', 'Даблмид', 'Впередплотва', 'Десперрок', 'Лаафект', 'Гуолан', 'Дэйлинс', 'Кайперс',
  'Демьяна', 'Дамнейшен', 'Золмар', 'Атейн', 'Келаний', 'Дамнейшн', 'Погода', 'Шадоудва'
};

-- 2597  Alterac Valley
-- 3358  Arathi Basin
-- 3820  Eye of the Storm
-- 4710  Isle of Conquest
-- 4384  Strand of the Ancients
-- 5449  The Battle for Gilneas
-- 5031  Twin Peaks
-- 3277  Warsong Gulch

local BATTLEFIELDS = {
    [30] = { name='Alterac Valley', size=40 },
   [489] = { name='Warsong Gulch', size=10 },
   [529] = { name='Arathi Basin', size=15 },
   [566] = { name='Eye of the Storm', size=15 },
   [607] = { name='Strand of the Ancients', size=15 },
   [628] = { name='Isle of Conquest', size=40 },
   [726] = { name='Twin Peaks', size=10 },
   [727] = { name='Silvershard Mines', size=10 },
   [761] = { name='The Battle for Gilneas', size=10 },
   [968] = { name='Eye of the Storm (RBG)', size=10 },
   [998] = { name='Temple of Kotmogu', size=10 },
  [1105] = { name='Deepwind Gorge', size=15 }
};

MrTarget = CreateFrame('Frame', 'MrTarget', UIParent);

function MrTarget:Load()
  self.loaded=true;
  self.active=false;
  self.version='v5.1.0';
  self.difficulty = false;
  self.frames={};
  self.player={};
  self.size=40;
  self.objectives=false;
  self.options_open=false;
  self:HelloWorld();
  self:GetOptions();
  self:Initialize();
  self:InitOptions();
end

function MrTarget:Initialize()
  if not self.frames.HARMFUL then
    self.frames.HARMFUL = MrTargetGroup:New('HARMFUL', false);
  end
  if not self.frames.HELPFUL then
    self.frames.HELPFUL = MrTargetGroup:New('HELPFUL', true);
  end
end

function MrTarget:Activate()
  if self:GetOption('ENEMY') then
    self.frames.HARMFUL:Activate();
  end
  if self:GetOption('FRIENDLY') then
    self.frames.HELPFUL:Activate();
  end
end

function MrTarget:ZoneChanged()
  for i=1, GetMaxBattlefieldID() do
    local mapId = select(8, GetInstanceInfo())
    if BATTLEFIELDS[mapId] then
      self.size = IsRatedBattleground() and 10 or BATTLEFIELDS[mapId].size;
      if self:GetOption('enabled') and not self.active then
        self.active = true;
        self:DisableOptions();
        self:ObjectivesFrame(true);
        self:UpdateOptions();
        return self:Activate();
      else
        return;
      end
    end
  end
  if not self.options_open then
    self.active = false;
    self:EnableOptions();
    self:ObjectivesFrame(false);
    self:Destroy();
  end
end

function MrTarget:ObjectivesFrame(active)
  if ObjectiveTrackerFrame then
    if active and not ObjectiveTrackerFrame.collapsed then
      ObjectiveTracker_Collapse();
      self.objectives = true;
    elseif not active and self.objectives and ObjectiveTrackerFrame.collapsed then
      ObjectiveTracker_Expand();
      self.objectives = false;
    end
  end
end

function MrTarget:PlayerLogin()
  if self.loaded then
    self.player.NAME = UnitName('player');
    self.player.CLASS = select(2, UnitClass('player'));
    self.player.FACTION = GetBattlefieldArenaFaction();
  end
end

function MrTarget:HelloWorld()
  local message = '|cFF00FFFF <MrTarget-'..self.version..'>|cFFFF0000 Even the Score.|r Type /mrt for interface options.';
  ChatFrame1:AddMessage(message, 0, 0, 0, GetChatTypeIndex('SYSTEM'));
end

function MrTarget:GetSize()
  return self.size;
end

function MrTarget:GetOptions()
  self.OPTIONS = MRTARGET_SETTINGS;
  if not self.OPTIONS or not self.OPTIONS[40] then
    self.OPTIONS = copy(DEFAULT_OPTIONS);
  end
end

function MrTarget:InitOptions()
  self.options_frame = CreateFrame('Frame', 'MrTargetOptions', UIParent);
  self.options_frame.name = GetAddOnMetadata('MrTarget', 'Title');
  self.options_frame.tabs = {};
  InterfaceOptions_AddCategory(self.options_frame);
  for k,v in pairs(BATTLEFIELD_SIZES) do
    self.options_frame.tabs[v] = CreateFrame('Frame', 'MrTargetOptions'..v, self.options_frame, 'MrTargetOptionsTemplate');
    self.options_frame.tabs[v].name = v..' man';
    self.options_frame.tabs[v].parent = self.options_frame.name;
    self.options_frame.tabs[v].size = v;
    self.options_frame.tabs[v].Title:SetText(string.upper(v..' Man Battlegrounds'));
    self.options_frame.tabs[v].Subtitle:SetText('MrTarget '..self.version);
    self.options_frame.tabs[v].okay = function() MrTarget:SaveOptions(); end;
    self.options_frame.tabs[v].default = function() MrTarget:DefaultOptions(); end;
    self.options_frame.tabs[v].cancel = function() MrTarget:CancelOptions(); end;
    self.options_frame.tabs[v]:SetScript('OnHide', function(self) MrTarget:CloseOptions(); end);
    self.options_frame.tabs[v]:SetScript('OnShow', function(self) MrTarget:OpenOptions(v); end);
    InterfaceOptions_AddCategory(self.options_frame.tabs[v]);
  end
  self:SetOptions(self.OPTIONS);
end

function MrTarget:InitNamingOptions(default)
  for k,v in pairs(BATTLEFIELD_SIZES) do
    UIDropDownMenu_Initialize(self.options_frame.tabs[v].Naming, function()
      for i, option in pairs({ 'Transmute', 'Transliterate', 'Ignore' }) do
        UIDropDownMenu_AddButton({ owner=self.options_frame.tabs[v].Naming, text=option, value=option, checked=nil, arg1=option, func=(function(_, value)
          UIDropDownMenu_ClearAll(self.options_frame.tabs[v].Naming);
          UIDropDownMenu_SetSelectedValue(self.options_frame.tabs[v].Naming, value);
          self:SetOption('naming', value);
        end)});
      end
    end);
    UIDropDownMenu_SetAnchor(self.options_frame.tabs[v].Naming, 16, 22, 'TOPLEFT', self.options_frame.tabs[v].Naming:GetName()..'Left', 'BOTTOMLEFT');
    UIDropDownMenu_JustifyText(self.options_frame.tabs[v].Naming, 'LEFT');
    UIDropDownMenu_SetSelectedValue(self.options_frame.tabs[v].Naming, default);
  end
end

function MrTarget:SetOptions(options)
  for k,v in pairs(BATTLEFIELD_SIZES) do
    self.options_frame.tabs[v].Enabled:SetChecked(options[v].ENABLED);
    self.options_frame.tabs[v].Enemy:SetChecked(options[v].ENEMY);
    self.options_frame.tabs[v].Friendly:SetChecked(options[v].FRIENDLY);
    self.options_frame.tabs[v].Power:SetChecked(options[v].POWER);
    self.options_frame.tabs[v].Range:SetChecked(options[v].RANGE);
    self.options_frame.tabs[v].Targeted:SetChecked(options[v].TARGETED);
    self.options_frame.tabs[v].Auras:SetChecked(options[v].AURAS);
    self.options_frame.tabs[v].Borderless:SetChecked(options[v].BORDERLESS);
    self.options_frame.tabs[v].Icons:SetChecked(options[v].ICONS);
    self.options_frame.tabs[v].Size:SetValue(options[v].SIZE);
    self.options_frame.tabs[v].Columns:SetValue(options[v].COLUMNS);
    self:InitNamingOptions(options[v].NAMING);
  end
  if self.size then
    self:SetOption('position', options[self.size].POSITION);
  end
end

function MrTarget:UpdateOptions()
  self:SetOptionSize(self:GetOption('SIZE'));
  self:SetOptionPosition(self:GetOption('POSITION'));
  self:SetOptionColumns(self:GetOption('COLUMNS'));
  self.frames.HARMFUL:SetMax(self.size);
  self.frames.HELPFUL:SetMax(self.size);
end

function MrTarget:SaveOptions()
  MRTARGET_SETTINGS = self.OPTIONS;
  for k,v in pairs(BATTLEFIELD_SIZES) do
    MRTARGET_SETTINGS[v].ENABLED = self.options_frame.tabs[v].Enabled:GetChecked();
    MRTARGET_SETTINGS[v].ENEMY = self.options_frame.tabs[v].Enemy:GetChecked();
    MRTARGET_SETTINGS[v].FRIENDLY = self.options_frame.tabs[v].Friendly:GetChecked();
    MRTARGET_SETTINGS[v].POWER = self.options_frame.tabs[v].Power:GetChecked();
    MRTARGET_SETTINGS[v].RANGE = self.options_frame.tabs[v].Range:GetChecked();
    MRTARGET_SETTINGS[v].TARGETED = self.options_frame.tabs[v].Targeted:GetChecked();
    MRTARGET_SETTINGS[v].AURAS = self.options_frame.tabs[v].Auras:GetChecked();
    MRTARGET_SETTINGS[v].BORDERLESS = self.options_frame.tabs[v].Borderless:GetChecked();
    MRTARGET_SETTINGS[v].ICONS = self.options_frame.tabs[v].Icons:GetChecked();
    MRTARGET_SETTINGS[v].SIZE = self.options_frame.tabs[v].Size:GetValue();
    MRTARGET_SETTINGS[v].COLUMNS = self.options_frame.tabs[v].Columns:GetValue();
  end
end

function MrTarget:CancelOptions()
  self:SetOptions(self.OPTIONS);
end

function MrTarget:DefaultOptions()
  self:SetOptions(DEFAULT_OPTIONS);
  self:OpenOptions(self.size);
end

function MrTarget:DisableOptions()
  for k,v in pairs(BATTLEFIELD_SIZES) do
    UIDropDownMenu_DisableDropDown(self.options_frame.tabs[v].Naming);
    self.options_frame.tabs[v].Enabled:Disable();
    self.options_frame.tabs[v].Enemy:Disable();
    self.options_frame.tabs[v].Friendly:Disable();
    self.options_frame.tabs[v].Power:Disable();
    self.options_frame.tabs[v].Auras:Disable();
    self.options_frame.tabs[v].Borderless:Disable();
    self.options_frame.tabs[v].Icons:Disable();
  end
end

function MrTarget:EnableOptions()
  for k,v in pairs(BATTLEFIELD_SIZES) do
    UIDropDownMenu_EnableDropDown(self.options_frame.tabs[v].Naming);
    self.options_frame.tabs[v].Enabled:Enable();
    self.options_frame.tabs[v].Enemy:Enable();
    self.options_frame.tabs[v].Friendly:Enable();
    self.options_frame.tabs[v].Power:Enable();
    self.options_frame.tabs[v].Borderless:Enable();
    self.options_frame.tabs[v].Icons:Enable();
    if self.OPTIONS.COLUMNS == 1 then
      self.options_frame.tabs[v].Auras:Enable();
    end
  end
end

function MrTarget:GetOption(option)
  return self.OPTIONS[self.size][string.upper(option)];
end

function MrTarget:SetOption(option, value)
  if self.size then
    self.OPTIONS[self.size][string.upper(option)] = value;
    if self.options_frame.tabs[self.size] and self.options_frame.tabs[self.size]:IsVisible() then
      self:OpenOptions(self.size);
    end
  end
end

function MrTarget:SetOptionColumns(columns)
  if self.frames.HELPFUL then
    self.frames.HELPFUL:SetColumns(columns);
  end
  if self.frames.HARMFUL then
    self.frames.HARMFUL:SetColumns(columns);
  end
  if self.options_frame.tabs[self.size] then
    if columns > 1 then
      self.options_frame.tabs[self.size].Auras:SetChecked(false);
      self.OPTIONS[self.size][string.upper('auras')] = false;
      self.options_frame.tabs[self.size].Auras:Disable();
    else
      self.options_frame.tabs[self.size].Auras:Enable();
    end
  end
end

function MrTarget:SetOptionPosition(position)
  self:SetOptionPositionHARMFUL(position.HARMFUL);
  self:SetOptionPositionHELPFUL(position.HELPFUL);
end

function MrTarget:SetOptionPositionHARMFUL(position)
  self.frames.HARMFUL.frame:ClearAllPoints();
  local ok, point, relativeTo, relativePoint, x, y = pcall(unpack, position);
  if not ok then point, relativeTo, relativePoint, x, y = unpack(DEFAULT_BATTLEFIELD_OPTIONS.POSITION.HARMFUL);
  end
  self.frames.HARMFUL.frame:SetPoint(point, relativeTo, relativePoint, x, y);
  self.frames.HARMFUL:UpdateOrientation();
end

function MrTarget:SetOptionPositionHELPFUL(position)
  self.frames.HELPFUL.frame:ClearAllPoints();
  local ok, point, relativeTo, relativePoint, x, y = pcall(unpack, position);
  if not ok then point, relativeTo, relativePoint, x, y = unpack(DEFAULT_BATTLEFIELD_OPTIONS.POSITION.HELPFUL);
  end
  self.frames.HELPFUL.frame:SetPoint(point, relativeTo, relativePoint, x, y);
  self.frames.HELPFUL:UpdateOrientation();
end

function MrTarget:SetOptionSize(size)
  if self.frames.HELPFUL then
    self.frames.HELPFUL.frame:SetScale(size/100);
  end
  if self.frames.HARMFUL then
    self.frames.HARMFUL.frame:SetScale(size/100);
  end
end

function MrTarget:OpenOptions(size)
  if not self.active then
    self.size = size;
    self:CloseOptions();
    self:Initialize();
    self.options_open = true;
    if self:GetOption('ENABLED') then
      self:ObjectivesFrame(true);
      if self:GetOption('ENEMY') then
         self.frames.HARMFUL:CreateStub(ENEMIES, size);
      else
        self.frames.HARMFUL:Destroy();
      end
      if self:GetOption('FRIENDLY') then
        self.frames.HELPFUL:CreateStub(FRIENDS, size);
      else
        self.frames.HELPFUL:Destroy();
      end
      if self:GetOption('BORDERLESS') then
        self.options_frame.tabs[size].Icons:Show();
      else
        self.options_frame.tabs[size].Icons:Hide();
      end
      self:UpdateOptions();
    else
      self:Destroy();
    end
  end
end

function MrTarget:CloseOptions()
  if not self.active then
    self:ObjectivesFrame(false);
    self.options_frame:Hide();
    self.options_open = false;
    self:Destroy();
  end
end

function MrTarget:Destroy()
  self.frames.HARMFUL:Destroy();
  self.frames.HELPFUL:Destroy();
end

function MrTarget:AddonLoaded(addon)
  if addon == 'MrTarget' then
    self:Load();
  end
end

function MrTarget:OnEvent(event, ...)
  if event == 'ADDON_LOADED' then self:AddonLoaded(...);
  elseif event == 'PLAYER_LOGIN' then self:PlayerLogin();
  elseif event == 'ACTIVE_TALENT_GROUP_CHANGED' then self:PlayerLogin();
  elseif event == 'PLAYER_ENTERING_WORLD' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED_NEW_AREA' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED_INDOORS' then self:ZoneChanged();
  end
end

MrTarget:SetScript('OnEvent', MrTarget.OnEvent);

MrTarget:RegisterEvent('PLAYER_ENTERING_WORLD');
MrTarget:RegisterEvent('ACTIVE_TALENT_GROUP_CHANGED');
MrTarget:RegisterEvent('ZONE_CHANGED');
MrTarget:RegisterEvent('ZONE_CHANGED_NEW_AREA');
MrTarget:RegisterEvent('ZONE_CHANGED_INDOORS');
MrTarget:RegisterEvent('PLAYER_LOGIN');
MrTarget:RegisterEvent('ADDON_LOADED');

SLASH_MRTARGET1 = '/mrt';
SLASH_MRTARGET2 = '/mrtarget';
function SlashCmdList.MRTARGET(cmd, box)
  if cmd == 'show' then
    MrTarget:Activate();
  elseif cmd == 'hide' then
    MrTarget:Destroy();
  else
    InterfaceOptionsFrame_OpenToCategory(MrTarget.options_frame);
    InterfaceOptionsFrame_OpenToCategory(MrTarget.options_frame.tabs[10]);
  end
end